### PR TITLE
Add lint to check for boolean comparison in assert macro calls

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This PR adds a lint to check if an assert macro is using a boolean as "comparison value". For example:

```rust
assert_eq!("a".is_empty(), false);
```

Could be rewritten as:

```rust
assert!(!"a".is_empty());
```

PS: The dev guidelines are amazing. Thanks a lot for writing them!

changelog: Add `bool_assert_comparison` lint

### CODE REVIEW

**Comment by camsteffen** [comment]
I'm pretty sure we are moving away from pre-expansion passes. Sorry that that means you need to re-work this to inspect expanded code. We have a util function `extract_assert_macro_args` to do this with the HIR. But it would be nice to implement this in an early pass (AST).

Tagging @rust-lang/clippy for input on this.

**Comment by GuillaumeGomez** [comment]
Makes sense. I'll rewrite it then.

**Comment by GuillaumeGomez** [comment]
Rewrite done. I'm very unhappy with the result though: if any of the macro gets updated, the lint will break. Also, the suggestion now is much worse since I can't simply show the macro arguments... Anyway, at least now it's fully running on early pass and doesn't require macro expansion anymore!

**Comment by Manishearth** [comment]
Yeah, that's fine, we have a fair number of lints that do that kind of thing, which is why we have all these desugaring functions in `utils`. We catch breakages when they happen and update the code; this isn't a huge deal since clippy is shipped with rustc

**Comment by GuillaumeGomez** [comment]
CI finally passed! \o/

File: clippy_lints/src/bool_assert_comparison.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                        }) = arg.kind
+                        {
+                            nb_bool_args += 1;
+                            bool_span = arg.span;
+                        }
+                    }
+                    #[allow(clippy::if_same_then_else)] // It allows better code reability here.
+                    if nb_bool_args > 1 {
+                        // Don't know what's going on here, but better leave things as is...
+                        return;
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:55]: I believe it could be `assert_eq!(a!(), b!())`. Should add a test for that and check `!arg.span.from_expansion()` (don't lint if the bool is from a macro).

File: clippy_lints/src/bool_assert_comparison.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    for arg in macro_args {
+                        if let ExprKind::Lit(Lit {
+                            kind: LitKind::Bool(_), ..
+                        }) = arg.kind
+                        {
+                            nb_bool_args += 1;
+                            bool_span = arg.span;
+                        }
+                    }
+                    #[allow(clippy::if_same_then_else)] // It allows better code reability here.
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:52]: I think it would be more readable to just check `nb_bool_args == 1` with whatever comments.

File: clippy_lints/src/bool_assert_comparison.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                        return;
+                    }
+
+                    span_lint_and_sugg(
+                        cx,
+                        BOOL_ASSERT_COMPARISON,
+                        bool_span,
+                        "assert macro with a boolean comparison",
+                        "replace it with",
+                        format!("{}!", &mac[..mac.len() - 3]),
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:67]: This is not intuitive. Add either a comment or a variable `let non_eq_mac = &mac[..mac.len() - 3];`.

File: clippy_lints/src/bool_assert_comparison.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                        Applicability::MachineApplicable,
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:68]: This is not machine-applicable unless you remove the bool arg and add `!` if the bool is `false`. Up to you if you want to implement that.

File: clippy_lints/src/bool_assert_comparison.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    } else if nb_bool_args == 0 {
+                        // No problem detected!
+                        return;
+                    }
+
+                    span_lint_and_sugg(
+                        cx,
+                        BOOL_ASSERT_COMPARISON,
+                        bool_span,
+                        "assert macro with a boolean comparison",
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:65]: The "literal" part is important.

```suggestion
                        format!("used {}! with a literal bool", mac),
```

File: clippy_utils/src/ast_utils.rs [hunk 6]
```diff
@@ -571,3 +572,36 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
...
+/// For example:
+/// `assert!(expr)` will return Some([expr])
+/// `debug_assert_eq!(a, b)` will return Some([a, b])
+pub fn extract_assert_macro_args(e: &Expr) -> Option<Vec<&P<Expr>>> {
+    fn get_args(e: &Expr) -> Option<Vec<&P<Expr>>> {
+        if_chain! {
+            if let ExprKind::Block(ref block, _) = e.kind;
+            if let StmtKind::Expr(ref expr) = block.stmts.get(0)?.kind;
+            if let ExprKind::Match(ref match_expr, _) = expr.kind;
+            if let ExprKind::Tup(ref tup) = match_expr.kind;
```
Comment (camsteffen) [inline clippy_utils/src/ast_utils.rs:589]: hint
```suggestion
            if let ExprKind::Tup([a, b]) = match_expr.kind;
```

File: clippy_utils/src/ast_utils.rs [hunk 7]
```diff
@@ -571,3 +572,36 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
...
 }
+
+/// Extract args from an assert-like macro.
+/// Currently working with:
+/// - `assert!`, `assert_eq!` and `assert_ne!`
+/// - `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!`
+/// For example:
+/// `assert!(expr)` will return Some([expr])
+/// `debug_assert_eq!(a, b)` will return Some([a, b])
+pub fn extract_assert_macro_args(e: &Expr) -> Option<Vec<&P<Expr>>> {
```
Comment (camsteffen) [inline clippy_utils/src/ast_utils.rs:583]: This function may as well determine which macro it is and return the name of the macro. You can call `is_direct_expn_of` as the final step.

You can return an array instead of `Vec`. You can deref the `P`.

```suggestion
pub fn extract_assert_macro_args(e: &Expr) -> Option<(&'static str, [&Expr; 2])> {
```

File: clippy_utils/src/ast_utils.rs [hunk 8]
```diff
@@ -571,3 +572,36 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
...
+                }).collect::<Vec<_>>());
+            }
+        }
+        None
+    }
+    if_chain! {
+        if let ExprKind::If(_, ref block, _) = e.kind;
+        if let StmtKind::Semi(ref e) = block.stmts.get(0)?.kind;
+        then {
+            return get_args(e);
```
Comment (camsteffen) [inline clippy_utils/src/ast_utils.rs:603]: You can easily avoid having an inner function like this.

```suggestion
            expr = e;
```

File: clippy_lints/src/bool_assert_comparison.rs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+declare_lint_pass!(BoolAssertComparison => [BOOL_ASSERT_COMPARISON]);
+
+impl EarlyLintPass for BoolAssertComparison {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &Expr) {
+        let macros = ["assert_eq", "debug_assert_eq"];
+        let inverted_macros = ["assert_ne", "debug_assert_ne"];
+
+        for mac in macros.iter().chain(inverted_macros.iter()) {
+            if is_direct_expn_of(e.span, mac).is_some() {
+                if let Some(macro_args) = ast_utils::extract_assert_macro_args(e) {
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:40]: There's a lot of indentation at this point making this hard to follow. But this may be addressed through my suggestions on `extract_assert_macro_args`.

**Review by camsteffen — CHANGES_REQUESTED** [review]
Not sure what the file `/bool_assert_eq` is. Probably need to remove.

File: clippy_lints/src/bool_assert_comparison.rs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+declare_lint_pass!(BoolAssertComparison => [BOOL_ASSERT_COMPARISON]);
+
+impl EarlyLintPass for BoolAssertComparison {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &Expr) {
+        let macros = ["assert_eq", "debug_assert_eq"];
+        let inverted_macros = ["assert_ne", "debug_assert_ne"];
+
+        for mac in macros.iter().chain(inverted_macros.iter()) {
+            if is_direct_expn_of(e.span, mac).is_some() {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:39]: Right, now that I don't show the full result anymore, it doesn't make much sense anymore...

File: clippy_lints/src/bool_assert_comparison.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:52]: I wanted to show both both cases so that it made code more easy to read, so as you prefer.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:55]: And both `a!()` and `b!()` would return a bool?

File: clippy_utils/src/ast_utils.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:583]: I'm not a big fan of doing multiple things at once. However I can return `Expr` directly.

File: clippy_utils/src/ast_utils.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:603]: Good idea!

File: clippy_lints/src/bool_assert_comparison.rs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    /// ```rust
+    /// // Bad
+    /// assert_eq!("a".is_empty(), false);
+    /// assert_ne!("a".is_empty(), true);
+    ///
+    /// // Good
+    /// assert!(!"a".is_empty());
+    /// ```
+    pub BOOL_ASSERT_COMPARISON,
+    pedantic,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:27]: I'm not sure it's about style here. It's really about using the invalid assert macro here.

File: clippy_lints/src/bool_assert_comparison.rs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    "Using a boolean as comparison value when there is no need"
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:28]: :+1:

File: clippy_lints/src/bool_assert_comparison.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:65]: :+1:

File: clippy_lints/src/bool_assert_comparison.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:68]: Not possible at all. In my first commit (when I was doing this before the macro expansion), it was possible but after the expansion it's just a huge mess. So I'll simply update the value.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:67]: :+1:

File: clippy_utils/src/ast_utils.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:589]: This is a good one, indeed. ;)

Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:589]: Ah no, doesn't work because it's a vec...

File: clippy_lints/src/bool_assert_comparison.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+                    let non_eq_mac = &mac[..mac.len() - 3];
+                    span_lint_and_sugg(
+                        cx,
+                        BOOL_ASSERT_COMPARISON,
+                        e.span,
+                        &format!("used `{}!` with a literal bool", mac),
+                        "replace it with",
+                        format!("{}!", non_eq_mac),
+                        Applicability::HasPlaceholders,
```
Comment (GuillaumeGomez) [inline clippy_lints/src/bool_assert_comparison.rs:68]: With this new `Applicability`, the suggestion doesn't seem to be displayed anymore. Weird...

File: clippy_lints/src/bool_assert_comparison.rs (hunk 1 above)
Comment (camsteffen) [reply clippy_lints/src/bool_assert_comparison.rs]: right

File: clippy_lints/src/bool_assert_comparison.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (camsteffen) [reply clippy_lints/src/bool_assert_comparison.rs:26]: It's not invalid. It accomplishes the same thing. Just bad style.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 4 above)
Comment (camsteffen) [reply clippy_lints/src/bool_assert_comparison.rs]: It is possible. Your suggestion would look like this:

```rust
format!("{}!({}{})", mac_name, not_op, arg_snippet)
```

where `not_op` is `"!"` or `""` and `arg_snippet` is the output of `snippet(arg_span)`.

File: clippy_utils/src/ast_utils.rs (hunk 7 above)
Comment (camsteffen) [reply clippy_utils/src/ast_utils.rs]: It's not multiple things at once if you think of it as "get the assert macro parts". We have many functions like this if you look around in this same module. If you like, you can return a struct to make the output more clear. Returning the macro name will make the util more useful in general.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 13 above)
Comment (camsteffen) [reply clippy_lints/src/bool_assert_comparison.rs:68]: Hmm try `MaybeIncorrect`? That's more appropriate anyways.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (camsteffen) [reply clippy_lints/src/bool_assert_comparison.rs:48]: What do you mean by "the full result"?

File: clippy_lints/src/bool_assert_comparison.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs]: Where do you get the arg span if you have more than 2 arguments? Because this is a nightmare to get arguments starting from the third.

As for `no_op`, take a look at my first commit. ;)

File: clippy_utils/src/ast_utils.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs]: I don't agree with you. We have the macro name already before even calling this function. There is no point into returning it from this function once again.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:68]: I tried all of the other values. None of them display the suggestion.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs]: Ok, gonna need to add a test for this then.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:48]: My suggestion doesn't show anymore what the `assert!` call should look like since that the `snippet` span returns the macro "definition" and not the arguments anymore.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:26]: Ok! Changing it to style then.

File: tests/ui/bool_assert_comparison.stderr [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  --> $DIR/bool_assert_comparison.rs:4:19
+   |
+LL |     () => { true },
+   |                   ^ no rules expected this token in macro call
+
+error: no rules expected the token `,`
+  --> $DIR/bool_assert_comparison.rs:7:19
+   |
+LL |     () => { true },
+   |                   ^ no rules expected this token in macro call
```
Comment (camsteffen) [inline tests/ui/bool_assert_comparison.stderr:11]: Not sure what is causing this error?

File: clippy_lints/src/bool_assert_comparison.rs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    ///
+    /// ```rust
+    /// // Bad
+    /// assert_eq!("a".is_empty(), false);
+    /// assert_ne!("a".is_empty(), true);
+    ///
+    /// // Good
+    /// assert!(!"a".is_empty());
+    /// ```
+    pub BOOL_ASSERT_COMPARISON,
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:25]: I think `bool_assert_eq` would be a good name, even though it includes `*_ne`, but whichever you choose is good.

File: clippy_utils/src/ast_utils.rs (hunk 7 above)
Comment (camsteffen) [reply clippy_utils/src/ast_utils.rs]: Well you wouldn't need to get the macro name before calling the function anymore. But it's okay to leave this as is if you prefer.

File: tests/ui/bool_assert_comparison.stderr (hunk 14 above)
Comment (GuillaumeGomez) [reply tests/ui/bool_assert_comparison.stderr:11]: Ah, my macro is badly formatted.

File: clippy_utils/src/ast_utils.rs [hunk 16]
```diff
@@ -571,3 +572,32 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
...
+        then {
+            expr = e;
+        }
+    }
+    if_chain! {
+        if let ExprKind::Block(ref block, _) = expr.kind;
+        if let StmtKind::Expr(ref expr) = block.stmts.get(0)?.kind;
+        if let ExprKind::Match(ref match_expr, _) = expr.kind;
+        if let ExprKind::Tup(ref tup) = match_expr.kind;
+        if let (Some(ref a), Some(ref b)) = (tup.get(0), tup.get(1));
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:596]:
```
error[E0529]: expected an array or slice, found `Vec<P<rustc_ast::Expr>>`
   --> clippy_utils/src/ast_utils.rs:596:16
    |
596 |         if let [a, b, ..] = &*tup;
    |                ^^^^^^^^^^ pattern cannot match with input type `Vec<P<rustc_ast::Expr>>`

error: aborting due to previous error
```

;)

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:596]: Went around with:

```rust
if let [a, b, ..] = tup.as_slice();
```

File: clippy_lints/src/bool_assert_comparison.rs (hunk 15 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:25]: I originally went for this name before realizing that there were `*_ne` macros. This is why I went for this name instead.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:68]: Still not working apparently... :-/

File: clippy_utils/src/ast_utils.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs]: Then I'll keep as is then.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:48]: It's still not good in case you have more than two arguments. Getting the arguments starting from the third is a nightmare... It was easy to do in the pre_expansion pass, but now it's just trying to go through the macro source code from a token tree.

File: clippy_lints/src/bool_assert_comparison.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs]: And done, no problem. :)

File: clippy_lints/src/bool_assert_comparison.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs]: Ah no nevermind, it's now displayed! Awesome, thanks for the fix!

**Comment by GuillaumeGomez** [comment]
Updated!

File: clippy_utils/src/ast_utils.rs [hunk 17]
```diff
@@ -571,3 +572,32 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
         _ => false,
     }
 }
+
+/// Extract args from an assert-like macro.
+/// Currently working with:
+/// - `assert!`, `assert_eq!` and `assert_ne!`
+/// - `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!`
+/// For example:
+/// `assert!(expr)` will return Some([expr])
```
Comment (camsteffen) [inline clippy_utils/src/ast_utils.rs:581]: This would not work for `assert!` anymore?

File: clippy_lints/src/bool_assert_comparison.rs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+                    }
+
+                    let non_eq_mac = &mac[..mac.len() - 3];
+                    span_lint_and_sugg(
+                        cx,
+                        BOOL_ASSERT_COMPARISON,
+                        span,
+                        &format!("used `{}!` with a literal bool", mac),
+                        "replace it with",
+                        format!("{}!", non_eq_mac),
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:67]: To match the span being linted

```suggestion
                        format!("{}!(..)", non_eq_mac),
```

**Review by camsteffen — CHANGES_REQUESTED** [review]
A couple minor changes and then LGTM! Please squash your commits.

File: clippy_utils/src/ast_utils.rs (hunk 17 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:581]: There is no reason it would not work for `assert!`, or did you mean something else?

Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:581]: Ah yes you're right! Hum... I'll update the function so it's not always returnins two arguments.

**Comment by GuillaumeGomez** [comment]
> A couple minor changes and then LGTM! Please squash your commits.

I'd rather keep the first commit if you don't mind. It was more complete (in its suggestion) and it can be used for potential future migrations if anyone needs to do the same (I might haha).

File: clippy_utils/src/ast_utils.rs (hunk 17 above)
Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:581]: Now it's returning a `vec`, not great but at least it works with `assert` too.

File: clippy_utils/src/ast_utils.rs [hunk 19]
```diff
@@ -571,3 +572,39 @@ pub fn eq_mac_args(l: &MacArgs, r: &MacArgs) -> bool {
...
+        if let StmtKind::Expr(ref expr) = block.stmts.get(0)?.kind;
+        if let ExprKind::Match(ref match_expr, _) = expr.kind;
+        if let ExprKind::Tup(ref tup) = match_expr.kind;
+        then {
+            if let [a, b, ..] = tup.as_slice() {
+                if let (&ExprKind::AddrOf(_, _, ref a), &ExprKind::AddrOf(_, _, ref b)) = (&a.kind, &b.kind) {
+                    return Some(vec![&*a, &*b]);
+                }
+            }
+            if let [a, ..] = tup.as_slice() {
```
Comment (camsteffen) [inline clippy_utils/src/ast_utils.rs:602]: Isn't the `, ..` redundant? It looks like this would be better written as a single `match &**tup { [a] => .., [a, b, ..] => .., }`.

Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:602]: It was in the case not both arguments were `AddrOf`, so we handle both in case there is only one argument and when there are two where the second is invalid for whatever reason.

**Comment by camsteffen** [comment]
> I'd rather keep the first commit if you don't mind.

No, I don't see how it would be useful. Pre-expansion passes are going away. Migrating to an early pass will always require re-thinking the implementation. I [added a comment](https://github.com/rust-lang/rust-clippy/pull/7090) to hopefully stop people short of wring pre-exapnsion passes in the future.

**Comment by GuillaumeGomez** [comment]
Makes me sad but ok. Just waiting for our last discussion to be resolved and then I'll squash.

File: clippy_utils/src/ast_utils.rs (hunk 19 above)
Comment (camsteffen) [reply clippy_utils/src/ast_utils.rs:602]: It could be problematic to return one arg when there is actually more. But if we're just working on hypotheticals, I would rather err on the side of simplicity.

Comment (GuillaumeGomez) [reply clippy_utils/src/ast_utils.rs:602]: Then I'll revert and explicitly not support `assert!`.

**Comment by GuillaumeGomez** [comment]
Removed the support of `assert!` in the function and squashed the commits.

**Comment by camsteffen** [comment]
Sorry just thought of one more case that I think is missing: `assert_eq!(a!(), expr)`. One arg is a macro. Should not lint.

**Comment by GuillaumeGomez** [comment]
`a()!` isn't evaluated as true or false even if that's the only thing in it. I'll add a test case though. ;)

**Comment by GuillaumeGomez** [comment]
Ah nevermind! It actually throws an error. Interesting...

File: clippy_lints/src/bool_assert_comparison.rs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+                    let nb_bool_args = is_bool_lit(a) as usize + is_bool_lit(b) as usize;
+
+                    if nb_bool_args != 1 {
+                        // If there are two boolean arguments, we definitely don't understand
+                        // what's going on, so better leave things as is...
+                        //
+                        // Or there is simply no boolean and then we can leave things as is!
+                        return;
+                    }
+                    if a.span.from_expansion() || b.span.from_expansion() {
```
Comment (camsteffen) [inline clippy_lints/src/bool_assert_comparison.rs:59]: You should allow the non-bool-literal to be from expansion.

Comment (GuillaumeGomez) [reply clippy_lints/src/bool_assert_comparison.rs:59]: Good point! I'm really lucky that you caught all those issues. Thanks!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
Is there anything else for me to do in here?

**Comment by camsteffen** [comment]
Nope! Thanks!

@bors r+

**Comment by GuillaumeGomez** [comment]
There is a failure because there is a missing "changelog" in the PR body but can't see anything about it in https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md. If someone has a link or a tip on what I need to update? :)

**Comment by camsteffen** [comment]
@bors retry

added a changelog line

**Comment by GuillaumeGomez** [comment]
> added a changelog line

Thank you!

**Comment by xFrednet** [comment]
Hey, was there a reason why the lint suggestion doesn't snip the non literal expression into the suggested code change? :upside_down_face:

**Comment by GuillaumeGomez** [comment]
If you have an improvement in mind, don't hesitate to send it or at least open an issue with your suggestion.
